homeassistant/components/amcrest/config_flow.py

```diff
@@ -0,0 +1,233 @@
+                    CONF_PASSWORD: user_input[CONF_PASSWORD],
+                }
+
+                return await self._create_entry()
+
+            except LoginError:
+                errors["base"] = "faulty_credentials"
+
+            except AmcrestError:
+                errors["base"] = "device_unavailable"
+
+        data = self.discovery_schema or {
+            vol.Required(CONF_NAME, default=DEFAULT_NAME): str,
+            vol.Required(CONF_HOST): str,
+            vol.Required(CONF_USERNAME): str,
+            vol.Required(CONF_PASSWORD): str,
+            vol.Required(CONF_PORT, default=DEFAULT_PORT): int,
+        }
+
+        return self.async_show_form(
+            step_id="user",
+            description_placeholders=self.device_config,
+            data_schema=vol.Schema(data),
+            errors=errors,
+        )
+
+    async def _create_entry(self):
+        """Create entry for device."""
+
+        return self.async_create_entry(
+            title=self.device_config[CONF_NAME], data=self.device_config
+        )
+
+
+class AmcrestOptionsFlowHandler(config_entries.OptionsFlow):
+    """Handle Amcrest device options."""
+
+    def __init__(self, config_entry):
+        """Initialize Amcrest device options flow."""
+        self.config_entry = config_entry
+        self.options = dict(config_entry.options)
+        self.device = None
+
+    async def async_step_init(self, user_input=None):
+        """Manage the Amcrest device options."""
+        self.device = self.hass.data[DOMAIN][DEVICES][self.config_entry.data[CONF_NAME]]
+        return await self.async_step_configure_stream()
+
+    async def async_step_configure_stream(self, user_input=None):
+        """Manage the Amcrest device options."""
+        if user_input is not None:
+            self.options.update(user_input)
+            return self.async_create_entry(title="", data=self.options)
+
+        return self.async_show_form(
+            step_id="configure_stream",
+            data_schema=vol.Schema(
+                {
+                    vol.Optional(
+                        CONF_STREAM_SOURCE,
+                        default=self.config_entry.options.get(
+                            CONF_STREAM_SOURCE, DEFAULT_STREAM_SOURCE
+                        ),
+                    ): vol.In(STREAM_SOURCE_LIST),
+                    vol.Optional(
+                        CONF_FFMPEG_ARGUMENTS,
+                        default=self.config_entry.options.get(
+                            CONF_FFMPEG_ARGUMENTS, DEFAULT_FFMPEG_ARGUMENTS
+                        ),
+                    ): str,
+                    vol.Optional(
+                        CONF_RESOLUTION,
+                        default=self.config_entry.options.get(
+                            CONF_RESOLUTION, DEFAULT_RESOLUTION
+                        ),
+                    ): vol.All(vol.In(RESOLUTION_LIST)),
+                    vol.Optional(
+                        CONF_AUTHENTICATION,
```

Comment:
This seems weird to allow configuring via options flow. If the auth is incorrect, it should open a config flow `reauth` step.

Reply:
not all devices / firmware use the same authentication protocol.

Reply:
It's not clear what the purpose of this authentication option is anymore.    In python-amcrest, the authentication to the camera first attempts to connect with Basic auth (used by very old Amcrest/Dahua devices), and failing that, it retries using Digest.    

Per the HA Amcrest doc,  you only use this option if the obsolete MJPEG stream option was selected, but then it says only Basic auth is supported by aiohttp.   So what good does specifying a different auth for MJPEG do if it's not supported?   If the original author of this code cannot answer this, I would suggest not including it in config flow going forward.

> **authentication** string (Optional, default: basic)
> Defines which authentication method to use only when stream_source is **mjpeg**. Currently, **aiohttp only support basic**.

Reply:
CONF_AUTHENTICATION is only used for MJPEG stream option
I think the reason for this option is that
the old models allowed to see the video stream without authentication, while the new models require basic authentication

Reply:
Yes, that makes sense it was added for newer cameras.   So in order for MJPEG with authentication other than basic, the code doesn't use **aiohttp** (http asyncio) and streams with some other method?   If not, then it doesn't make sense if in fact aiohttp doesn't support anything other than basic.

Reply:
aiohttp works with basic authentication when authentication is required, but works even if authentication is not required (none)

Reply:
Yes, I get that.  Basic or no auth is acceptable.    My question is what is the purpose of this option for specifying digest or whatever else other than basic if it's not supported?    I just wanted to see if config could be simplified a little bit by not incorporating an option that may have no practical use.  It's no big deal - so nevermind!

Reply:
with that option you can specify "basic" authentication or "none" 
if u have a newer camera select "basic"
if u have old model camera select none

I don't understand how it is possible to remove the option and at the same time be able to maintain compatibility with both new and old cameras

Reply:
I fully understand the purpose of this option now.  Thanks for clarifying.   Yes, it needs to be retained for backward  compatibility.  

If in the config UI "CONF_Authentication" was only available as an option if MJPEG stream was selected, it would be less confusing.

Reply:
I don't think config flow offers these possibilities
But if there are, someone can you point me to a manual?